src/profile-logic/call-node-info.js

```diff
@@ -278,51 +274,507 @@ export class CallNodeInfoImpl implements CallNodeInfo {
+// The number of roots is the same as the number of functions in the funcTable.
+type InvertedRootCallNodeTable = {|
+  category: Int32Array, // IndexIntoFuncTable -> IndexIntoCategoryList
+  subcategory: Int32Array, // IndexIntoFuncTable -> IndexIntoSubcategoryListForCategory
+  innerWindowID: Float64Array, // IndexIntoFuncTable -> InnerWindowID
+  // IndexIntoNativeSymbolTable: all frames that collapsed into this call node inlined into the same native symbol
+  // -1: divergent: some, but not all, frames that collapsed into this call node were inlined, or they are from different symbols
+  // null: no inlining
+  sourceFramesInlinedIntoSymbol: Array<IndexIntoNativeSymbolTable | -1 | null>,
+  // The (exclusive) end of the suffix order index range for each root node.
+  // The beginning of the range is given by suffixOrderIndexRangeEnd[i - 1], or by
+  // zero. This is possible because both the inverted root order and the suffix order
+  // are determined by the func order.
+  suffixOrderIndexRangeEnd: Uint32Array, // IndexIntoFuncTable -> SuffixOrderIndex,
+  length: number,
+|};
+
+// Information about the non-root nodes of the inverted call tree. This table
+// grows on-demand, as new inverted call nodes are materialized.
+type InvertedNonRootCallNodeTable = {|
+  prefix: InvertedCallNodeHandle[],
+  func: IndexIntoFuncTable[], // IndexIntoInvertedNonRootCallNodeTable -> IndexIntoFuncTable
+  pathHash: string[], // IndexIntoInvertedNonRootCallNodeTable -> string
+  category: IndexIntoCategoryList[], // IndexIntoInvertedNonRootCallNodeTable -> IndexIntoCategoryList
+  subcategory: IndexIntoSubcategoryListForCategory[], // IndexIntoInvertedNonRootCallNodeTable -> IndexIntoSubcategoryListForCategory
+  innerWindowID: InnerWindowID[], // IndexIntoInvertedNonRootCallNodeTable -> InnerWindowID
+  // IndexIntoNativeSymbolTable: all frames that collapsed into this call node inlined into the same native symbol
+  // -1: divergent: some, but not all, frames that collapsed into this call node were inlined, or they are from different symbols
+  // null: no inlining
+  sourceFramesInlinedIntoSymbol: Array<IndexIntoNativeSymbolTable | -1 | null>,
+  suffixOrderIndexRangeStart: SuffixOrderIndex[], // IndexIntoInvertedNonRootCallNodeTable -> SuffixOrderIndex
+  suffixOrderIndexRangeEnd: SuffixOrderIndex[], // IndexIntoInvertedNonRootCallNodeTable -> SuffixOrderIndex
+
+  // Non-null for non-root nodes whose children haven't been created yet.
+  // For a non-root node x of the inverted tree, let k = depth[x] its depth in the inverted tree,
+  // and deepNodes = deepNodes[x] be its non-null deep nodes.
+  // Then, for every index i in suffixOrderIndexRangeStart[x]..suffixOrderIndexRangeEnd[x],
+  // the k'th prefix node of suffixOrderedCallNodes[i] is stored at deepNodes[x][i - suffixOrderIndexRangeStart[x]].
+  deepNodes: Array<Uint32Array | null>, // IndexIntoInvertedNonRootCallNodeTable -> (Uint32Array | null)
```

Comment:
I wonder of this could be `deepNodesCache`, to better refer that it's used as an optimizing trick... but up to you

Reply:
(Not asking to change the name now, this can be done in a follow-up if we feel it's better)

Another idea is that instead of the terminology of "deep node", it could be "ancestor node". Indeed what I struggled with is that the values in the Uint32Array are call node indexes in the non inverted call tree, they do not represent call nodes in the inverted call tree. But the term "deep" implies to me that they are children of this inverted call node (they're not!).
But I'm not sure of that so it might be good to wait (also I'd prefer to see this landed rather nitpicking on the exact terms)

Reply:
I like the name `deepNodesCache`, happy to rename it to this in a follow-up.

Reply:
I was thinking about the term "ancestor node" too but I was worried that it would be to easy to think it was referring to an ancestor in the *inverted* tree.